cmd/syncthing/main.go

```diff
@@ -568,6 +588,46 @@ func syncthingMain(runtimeOptions RuntimeOptions) {
 		os.Exit(syncthing.ExitError.AsInt())
 	}
 
+	// Candidate builds should auto upgrade. Make sure the option is set,
+	// unless we are in a build where it's disabled or the STNOUPGRADE
+	// environment variable is set.
+
+	if build.IsCandidate && !upgrade.DisabledByCompilation && !runtimeOptions.NoUpgrade {
+		l.Infoln("Automatic upgrade is always enabled for candidate releases.")
+		if opts := cfg.Options(); opts.AutoUpgradeIntervalH == 0 || opts.AutoUpgradeIntervalH > 24 {
+			opts.AutoUpgradeIntervalH = 12
+			// Set the option into the config as well, as the auto upgrade
+			// loop expects to read a valid interval from there.
+			cfg.SetOptions(opts)
+			cfg.Save()
+		}
+		// We don't tweak the user's choice of upgrading to pre-releases or
+		// not, as otherwise they cannot step off the candidate channel.
+	}
+
+	shouldAutoUpgrade := false
+
+	if opts := cfg.Options(); opts.AutoUpgradeIntervalH > 0 {
+		if runtimeOptions.NoUpgrade {
+			l.Infof("No automatic upgrades; STNOUPGRADE environment variable defined.")
+		} else {
+			release, err := checkUpgrade()
+			if err == nil {
+				if err = performUpgradeDirect(release); err == nil {
+					l.Infof("Upgraded to %q, exiting now.", release.Tag)
+					os.Exit(syncthing.ExitUpgrade.AsInt())
+				}
+			}
+			if err != upgrade.ErrUpgradeUnsupported {
+				shouldAutoUpgrade = true
+			} else if err != nil {
+				if _, ok := err.(errNoUpgrade); !ok {
```

Comment:
This branch will only be taken if err == ErrUpgradeUnsupported, or my double negation logic parser is bad (in which case this also needs simplification I think, as I've stared at these if clauses now for like half a minute)

Reply:
Thanks for the catch. And I refactored it, should be somewhat less convoluted now (still quite a few ifs though).